### PR TITLE
Introduce bats as successor to the home-grown integration test framework

### PR DESCRIPTION
**Pull request by tobim**
Moving away from the homegrown integration testing tool to a test framework that works directly in the shell environment offers a big increase of flexibility at a rather modest cost in verbosity.

This solves 2 long-standing pain points:
- it is possible to test concurrent client processes. See `database.bats:parallel clients` for an example.
- Disk- and IPC Message format stability can be tested by running specific older versions of VAST via docker or nix directly inside of a test.

The overhead of the test runner itself is negligible compared to the commands that regularly executed as part of the tests. I ported the "Server Zeek conn log" integration test (now called "batch sizes") over, and both frameworks indeed show comparable execution times.

There is one drawback that needs to be mentioned: `bats` has no functionality to update expectation values like the integration frameworks `--update` flag. It may be possible to retrofit this features with a custom wrapper for the `assert_output` function.

Fixes tenzir/issues#20

### CODE REVIEW

File: cmake/VASTRegisterPlugin.cmake [hunk 1]
```diff
@@ -466,6 +466,44 @@ function (VASTRegisterPlugin)
...
+      \"\\
+      #!/bin/sh
+
+      export TENZIR_PLUGIN_DIRS=\\\${TENZIR_PLUGIN_DIRS:+\\\${TENZIR_PLUGIN_DIRS}:}${PLUGIN_OUTPUT_DIRECTORY}
+      exec \\\"${TENZIR_BINARY_DIR}/\${EXECUTABLE}\\\" \\\"\\\$@\\\"\")
+      file(CHMOD \"${CMAKE_CURRENT_BINARY_DIR}/bin/\${EXECUTABLE}\"
+        FILE_PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE
+                         GROUP_READ GROUP_WRITE GROUP_EXECUTE)")
+    add_custom_target(
+      ${PLUGIN_TARGET}_wrapper ALL
```
Comment (dominiklohmann) [inline cmake/VASTRegisterPlugin.cmake:492]: Nit: targets with dashes and variables with underscores for consistency.

File: vast/CMakeLists.txt [hunk 2]
```diff
@@ -87,6 +87,39 @@ elseif (${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
...
+  cli-test-core
+  COMMAND
+    ${CMAKE_COMMAND} -E env
+    PATH="$<TARGET_FILE_DIR:vast::tenzir>:$ENV{PATH}:${CMAKE_CURRENT_SOURCE_DIR}/cli-test/bats/bin"
+    bats "-T" "${CMAKE_CURRENT_SOURCE_DIR}/cli-test/tests"
+  COMMENT "Executing CLI tests..."
+  USES_TERMINAL)
+
+add_dependencies(cli-test-core tenzir)
+add_dependencies(cli-test cli-test-core)
```
Comment (dominiklohmann) [inline vast/CMakeLists.txt:104]: More of a question than a request: Do you think it'd be possible to call `add_test` for every individual test suite? That way, we wouldn't need the separate target at all.

**Review by dominiklohmann — COMMENTED** [review]
A first round of comments after playing around with it a bit locally. This is fantastic, but I've got a few problems locally still.

File: vast/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (tobim) [reply vast/CMakeLists.txt:104]: That seems to happen in CI as well, good to know that you can reproduce.

Comment (tobim) [reply vast/CMakeLists.txt:104]: I don't think this is trivial, so I would like to avoid it in the context of this PR. Maybe as a follow up.

[the remark this answers is not in the document]
Comment (tobim) [reply vast/CMakeLists.txt]: CI is now working, can you try again?

Comment (tobim) [reply vast/CMakeLists.txt]: I don't think the value-for-money of this integration is high.
TBH I think it would be a better idea to just call `add_test` for each unit test binary and each functional test folder.
Let's discuss the pros and cons before jumping into action.

File: libvast/src/plugin.cpp [hunk 3]
```diff
@@ -176,10 +176,15 @@ caf::expected<detail::stable_set<std::string>> resolve_plugin_names(
...
+  auto path_map = detail::stable_map<std::string, std::string>{};
+  for (auto& path : paths_or_names) {
+    path_map.insert({std::filesystem::path{path}.filename(), std::move(path)});
+  }
   auto paths = detail::stable_set<std::string>{};
-  paths.insert(std::make_move_iterator(paths_or_names.begin()),
-               std::make_move_iterator(paths_or_names.end()));
+  for (auto& [_, path] : path_map) {
+    paths.insert(std::move(path));
+  }
```
[the remark this answers is not in the document]
Comment (tobim) [reply libvast/src/plugin.cpp]: This came up while testing the integration for standalone plugins. Left an explanation in the code.

File: plugins/web/functional-test/tests/serve.bats [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  # The next set should be the remaining 4 results.
+  check jq -nec "${second_result} | .events | length"
+
+  # Pull once more if necessary.
+  if jq -nec "${continuation_token} != null"; then
+    third_result="$(curl -XPOST -H "Content-Type: application/json" -d "{\"serve_id\": \"version\", \"continuation_token\": \"${continuation_token}\", \"timeout\": \"5s\", \"max_events\": 1}" http://127.0.0.1:5160/api/v0/serve)"
+    assert jq -nec "${third_result} | .next_continuation_token == null"
+    assert jq -nec "(${third_result} | .events | length) == 0"
+  fi
+}
```
Comment (tobim) [inline plugins/web/functional-test/tests/serve.bats:51]: This is now more restrictive than the old test, which calls serve in a loop until the continuation token is null. So far this test seems to be stable, but I'm not sure if the assumption that the token will be null with either the last events of directly afterwards.

File: vast/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (dominiklohmann) [reply vast/CMakeLists.txt]: Sadly this does still not quite work under macOS. There are some complaints about my locale, and there is a non-portable use `realpath` in `check.bash`.

```
❯ cmake --build build --target functional-test
[0/2] Re-checking globbed directories...
[11/13] Executing functional tests...
bash: warning: setlocale: LC_ALL: cannot change locale (en_US@rg=dezzzz-u-hc-h23-u-ca-gregory-u-nu-latn): Invalid argument
bash: warning: setlocale: LC_ALL: cannot change locale (en_US@rg=dezzzz-u-hc-h23-u-ca-gregory-u-nu-latn): Invalid argument
bash: warning: setlocale: LC_ALL: cannot change locale (en_US@rg=dezzzz-u-hc-h23-u-ca-gregory-u-nu-latn): Invalid argument
bash: warning: setlocale: LC_ALL: cannot change locale (en_US@rg=dezzzz-u-hc-h23-u-ca-gregory-u-nu-latn): Invalid argument
bash: warning: setlocale: LC_ALL: cannot change locale (en_US@rg=dezzzz-u-hc-h23-u-ca-gregory-u-nu-latn): Invalid argument
bash: warning: setlocale: LC_ALL: cannot change locale (en_US@rg=dezzzz-u-hc-h23-u-ca-gregory-u-nu-latn): Invalid argument
bash: warning: setlocale: LC_ALL: cannot change locale (en_US@rg=dezzzz-u-hc-h23-u-ca-gregory-u-nu-latn): Invalid argument
bash: warning: setlocale: LC_ALL: cannot change locale (en_US@rg=dezzzz-u-hc-h23-u-ca-gregory-u-nu-latn): Invalid argument
bash: warning: setlocale: LC_ALL: cannot change locale (en_US@rg=dezzzz-u-hc-h23-u-ca-gregory-u-nu-latn): Invalid argument
database.bats
 bash: warning: setlocale: LC_ALL: cannot change locale (en_US@rg=dezzzz-u-hc-h23-u-ca-gregory-u-nu-latn): Invalid argument                                                                             1/4 in 1 sec
 ✗ import and export commands [640]
   (from function `compare_or_update' in file /Users/dominiklohmann/Desktop/vast/bats/tenzir/functional-test/tests/..//bats-tenzir/src/check.bash, line 48,
    from function `check' in file /Users/dominiklohmann/Desktop/vast/bats/tenzir/functional-test/tests/..//bats-tenzir/src/check.bash, line 89,
    in test file /Users/dominiklohmann/Desktop/vast/bats/tenzir/functional-test/tests/database.bats, line 27)
     `< "$DATADIR/suricata/eve.json" \' failed
   TENZIR_AUTOMATIC_REBUILD=0
   TENZIR_BARE=1
   TENZIR_PLUGINS=
         _____ _____ _   _ ________ ____
        |_   _| ____| \ | |__  /_ _|  _ \
          | | |  _| |  \| | / / | || |_) |
          | | | |___| |\  |/ /_ | ||  _ <
          |_| |_____|_| \_/____|___|_| \_\

         v4.0.0-rc8-180-gc4e92e62bb-dirty
   Visit https://app.tenzir.com to get started.

   [21:27:54.177] loaded configuration file: "/Users/dominiklohmann/.config/tenzir/tenzir.yaml"
   [21:27:54.334] node is listening on 127.0.0.1:49966
   [21:27:54.602] importer adds suricata-reader source
   [21:27:54.605] importer removes suricata-reader source
   realpath: illegal option -- -
   usage: realpath [-q] [path ...]
initiating graceful shutdown... (repeat request to terminate immediately)
   Terminated: 15
```

[the remark this answers is not in the document]
Comment (tobim) [reply vast/CMakeLists.txt]: Pushed a potential fix for the `realpath` issue. I think I need some help with the locale because I don't know where that is coming from exactly.

[the remark this answers is not in the document]
Comment (dominiklohmann) [reply vast/CMakeLists.txt]: The `realpath` issue is fixed, but the tests stall for me with no explanation at the parallel import test, and the `setlocale` issue is very noisy.

Might make sense to take a look at this together when we're in the office the next time, I'm not really sure where to start debugging this.

**Comment by mavam** [comment]
After fixing the `zcat` madness, what's left is improving the shell quoting issues with `jq`. Almost there!

File: vast/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (mavam) [reply vast/CMakeLists.txt]: I think this was the `zcat` bug that is now fixed.

**Comment by mavam** [comment]
@tobim CI bug or something deeper?

**Comment by mavam** [comment]
When I run this locally, I get no failures:

```
❯ cmake --build build --target functional-test -j1
[0/2] Re-checking globbed directories...
[11/13] Executing functional tests...
 [setup_suite.bash, line 1]
 setup_suite >> "$BATS_OUT" 2>&1
 bats_require_minimum_version 1.8.0
 bats_load_library bats-tenzir
$$$$ [load.bash, line 6]
$$$$ source "$(dirname "${BASH_SOURCE[0]}")/src/debug.bash"
$$$$ dirname "${BASH_SOURCE[0]}"
$$$$ source "$(dirname "${BASH_SOURCE[0]}")/src/assert_sorted.bash"
$$$$ dirname "${BASH_SOURCE[0]}"
$$$$ source "$(dirname "${BASH_SOURCE[0]}")/src/check.bash"
$$$$ dirname "${BASH_SOURCE[0]}"
$$$$ source "$(dirname "${BASH_SOURCE[0]}")/src/fixtures.bash"
$$$$ dirname "${BASH_SOURCE[0]}"
 [setup_suite.bash, line 5]
 setup_state_dir
$ [fixtures.bash, line 41]
$ mkdir -p tenzir-functional-test-state
database.bats
 ✓ import and export commands
$$ [database.bats, line 21]
$$ teardown >> "$BATS_OUT" 2>&1
$$ teardown_node
$$$ [fixtures.bash, line 24]
$$$ kill "$NODE_PID"
$$$ local seconds=5
$$$ local killerPid=$!
$$$ wait "$NODE_PID"
$$$ sleep ${seconds}
$$$ sync
$$$ pkill -P "$killerPid"
$$$ exec {NODE_OUT}>&-
$$ [database.bats, line 23]
$$ teardown_db
$$$ [fixtures.bash, line 9]
$$$ [ -n "$BATS_TEST_COMPLETED" ]
$$$ [ "$BATS_TEST_COMPLETED" -eq 1 ]
$$$ rm -rf "$TENZIR_DB_DIRECTORY"
 ✓ parallel imports
$$ [database.bats, line 21]
$$ teardown >> "$BATS_OUT" 2>&1
$$ teardown_node
$$$ [fixtures.bash, line 24]
$$$ kill "$NODE_PID"
$$$ local seconds=5
$$$ local killerPid=$!
$$$ sleep ${seconds}
$$$ wait "$NODE_PID"
$$$ sync
$$$ pkill -P "$killerPid"
$$$ exec {NODE_OUT}>&-
$$ [database.bats, line 23]
$$ teardown_db
$$$ [fixtures.bash, line 9]
$$$ [ -n "$BATS_TEST_COMPLETED" ]
$$$ [ "$BATS_TEST_COMPLETED" -eq 1 ]
$$$ rm -rf "$TENZIR_DB_DIRECTORY"
 ✓ batch size
$$ [database.bats, line 21]
$$ teardown >> "$BATS_OUT" 2>&1
$$ teardown_node
$$$ [fixtures.bash, line 24]
$$$ kill "$NODE_PID"
$$$ local seconds=5
$$$ local killerPid=$!
$$$ wait "$NODE_PID"
$$$ sleep ${seconds}
$$$ sync
$$$ pkill -P "$killerPid"
$$$ exec {NODE_OUT}>&-
$$ [database.bats, line 23]
$$ teardown_db
$$$ [fixtures.bash, line 9]
$$$ [ -n "$BATS_TEST_COMPLETED" ]
$$$ [ "$BATS_TEST_COMPLETED" -eq 1 ]
$$$ rm -rf "$TENZIR_DB_DIRECTORY"
version.bats
 ✓ the version command
$ teardown_suite >> "$BATS_OUT" 2>&1
$ bats_load_library bats-tenzir
$$$$$ [load.bash, line 6]
$$$$$ source "$(dirname "${BASH_SOURCE[0]}")/src/debug.bash"
$$$$$ dirname "${BASH_SOURCE[0]}"
$$$$$ source "$(dirname "${BASH_SOURCE[0]}")/src/assert_sorted.bash"
$$$$$ dirname "${BASH_SOURCE[0]}"
$$$$$ source "$(dirname "${BASH_SOURCE[0]}")/src/check.bash"
$$$$$ dirname "${BASH_SOURCE[0]}"
$$$$$ source "$(dirname "${BASH_SOURCE[0]}")/src/fixtures.bash"
$$$$$ dirname "${BASH_SOURCE[0]}"
$ [setup_suite.bash, line 10]
$ try_remove_state_dir
$$ [fixtures.bash, line 46]
$$ [ "$(ls -A tenzir-functional-test-state)" == "" ]
$$ ls -A tenzir-functional-test-state
$$ rmdir tenzir-functional-test-state

4 tests, 0 failures

[12/13] Executing web functional tests...
/usr/lib/bats:/Users/mavam/code/tenzir/tenzir/bats/build/bin/../share/tenzir/functional-test
rest.bats
 ✓ status endpoint [0]
 ✓ authenticated endpoints [1000]
 ✓ TLS endpoints [1000]
serve.bats
 ✓ serve endpoint [3000]

4 tests, 0 failures in 5 seconds
```

**Comment by mavam** [comment]
@tobim @dominiklohmann one thought: do you remember that I needed to update Bash on my Mac to get things going? I'm now on:

```
❯ bash --version
GNU bash, version 5.2.15(1)-release (aarch64-apple-darwin22.1.0)
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
❯ which bash
/opt/homebrew/bin/bash
```

I was on the macOS bash, which is stuck in 3.x:

```
❯ /bin/bash --version
GNU bash, version 3.2.57(1)-release (arm64-apple-darwin22)
Copyright (C) 2007 Free Software Foundation, Inc.
```

I can't find any `brew install bash` in our CI. Is that perhaps the problem?

File: scripts/macOS/install-dev-dependencies.sh
Comment (dominiklohmann) [inline scripts/macOS/install-dev-dependencies.sh:1]: The bash 4+ and bats requirements should make it into the build from source guide as well, I think.

File: .github/workflows/style-check.yaml [hunk 5]
```diff
@@ -51,6 +51,11 @@ jobs:
             xargs -0 --no-run-if-empty black || true
           git diff --exit-code
 
+      - name: ShellCheck
+        run: |
+          git diff --name-only --no-color --diff-filter=ACM $(git merge-base origin/master HEAD) |
```
Comment (dominiklohmann) [inline .github/workflows/style-check.yaml:56]:
```suggestion
          git diff --name-only --no-color --diff-filter=ACM $(git merge-base origin/main HEAD) |
```

**Review by dominiklohmann — APPROVED** [review]
I think we have loads to migrate and a lot of small improvements we can make for this, but as this now runs through portably I think we should ship it so we can all write our new tests with this new framework. 

Thank you so much for pushing this forward! Not sure when we first started talking about this, but it's probably been two years at this point. ❤️

**Comment by tobim** [comment]
> I can't find any brew install bash in our CI. Is that perhaps the problem?

FTR, that was indeed one of the problems. The other one was that I relied on `realpath`, which is not in a default macOS install before 13.0.
